File: src/main/java/eu/karelhovorka/zpevnik/text/Section.java
package eu.karelhovorka.zpevnik.text;


import static eu.karelhovorka.zpevnik.util.Preconditions.checkNotNull;

public class Section {

    private final String content;

    private final String[] lines;

    private final SectionType type;

    private final Section copyOf;

    private final int typeIndex;

    public Section(String content, SectionType type) {
        this(content, type, null, 0);
    }

    public Section(String content, SectionType type, Section copyOf, int typeIndex) {
        checkNotNull(content);
        checkNotNull(type);
        this.content = content;
        this.type = type;
        this.copyOf = copyOf;
        this.typeIndex = typeIndex;
        this.lines = content.split("\n");
    }

    public String getContent() {
        return content;
    }

    public SectionType getType() {
        return type;
    }

    public boolean isCopy() {
        return copyOf != null;
    }

    public boolean sameContentAndType(Section another) {
        if (another == null) {
            return false;
        }
        if (type != another.type) {
            return false;
        }
        if (content.equals(another.getContent())) {
            return true;
        }
        if (content.equals("")) {
            return true;
        }
        if (content.endsWith("...") && another.content.startsWith(content.substring(0, content.length() - 3))) {
            return true;
        }
        return false;
    }

    public String getCss() {
        StringBuilder sb = new StringBuilder();
        sb.append(" section");
        sb.append(" section-type-" + type.css());
        sb.append(" section-type-" + type.css() + typeIndex);
        return sb.toString();
    }

    public String getShortName() {
        return type.getShortcut(typeIndex);
    }

    public Section getCopyOf() {
        return copyOf;
    }

    public int getTypeIndex() {
        return typeIndex;
    }

    public static Section makeCopyOf(Section previousSection) {
        return new Section(previousSection.content, previousSection.type, previousSection, previousSection.typeIndex + 1);
    }

    @Override
    public String toString() {
        return "Section{" +
                "type='" + type + '\'' +
                ", typeIndex=" + typeIndex +
                '}';
    }

    public String[] getLines() {
        return lines;
    }
}

File: src/main/java/eu/karelhovorka/zpevnik/text/Section.kt
package eu.karelhovorka.zpevnik.text


import eu.karelhovorka.zpevnik.text.section.SectionText
import java.lang.StringBuilder

class Section @JvmOverloads constructor(val content: SectionText, val type: ISectionType = SectionType.UNKNOWN, val copyOf: Section? = null, val copyIndex: Int = 0, val index: Int = 0) {

    val isCopy: Boolean
        get() = copyOf != null

    val css: String
        get() {
            val sb = StringBuilder()
            sb.append(" section")
            sb.append(" section-type-" + type.css())
            sb.append(" section-type-" + type.css() + index)
            return sb.toString()
        }

    val shortName: String
        get() = type.getShortcut(index)

    val longName: String
        get() = type.getLongName(index)


    fun sameContentAndType(another: Section?): Boolean {
        if (another == null) {
            return false
        }
        if (type != another.type) {
            return false
        }
        if (content == another.content) {
            return true
        }
        if (index == another.index) {
            return true
        }
        return false
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Section

        if (content != other.content) return false
        if (type != other.type) return false
        if (copyOf != other.copyOf) return false
        if (copyIndex != other.copyIndex) return false
        if (index != other.index) return false

        return true
    }

    override fun hashCode(): Int {
        var result = content.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + (copyOf?.hashCode() ?: 0)
        result = 31 * result + copyIndex
        result = 31 * result + index
        return result
    }

    override fun toString(): String {
        return "Section(content=$content, type=$type, copyOf=$copyOf, copyIndex=$copyIndex, index=$index)"
    }

    companion object {

        fun makeCopyOf(previousSection: Section): Section {
            return Section(previousSection.content, previousSection.type, null, previousSection.copyIndex + 1, previousSection.index)
        }
    }
}
